open-apparel-registry/open-apparel-registry
# Add ApiLimit and ApiBlock

## Overview

Adds models to support metering API access. The ApiLimit describes the
number of Api requests contributors can make monthly. The ApiBlock
records times when the allowed amount has been exceeded and provides
a location for recording offered grace periods.

Connects #1138 

## Demo

ApiBlock
<img width="763" alt="Screen Shot 2020-10-19 at 12 32 27 PM" src="https://user-images.githubusercontent.com/21046714/96479668-449ad180-1207-11eb-9ce8-424a2006abab.png">
ApiLimit
<img width="453" alt="Screen Shot 2020-10-19 at 12 33 08 PM" src="https://user-images.githubusercontent.com/21046714/96479672-46649500-1207-11eb-8e9a-6de45fb0318c.png">

## Testing Instructions

* Checkout this branch
* Run `./scripts/update`
* Use `./scripts/manage shell_plus` to create a new api limit and api block and to confirm the block was added
```
ApiLimit.objects.create(contributor_id=2, monthly_limit=50)
block = ApiBlock.objects.create(contributor_id=2, limit=50, actual=100, grace_limit=100, grace_reason='testing123',until='2020-10-30')
block.__dict__
```
* Navigate to [http://localhost:8081/admin/api/apilimit](http://localhost:8081/admin/api/apilimit) and confirm the limit exists
* Select the new limit and confirm that the monthly_limit can be updated but the contributor is readonly.

## Checklist
